PB-1112: Use pytest

msom opened the pull request:
This PR improves the test suite, which previously contained a mix of Django and pytest tests. It

- moves all tests to `app/tests` to align with the structure of the other services
- removes all `__init__.py` files within test folders to prevent test code from being auto-discovered outside of pytest
- transfers the functionality of `utils/testing.py` to fixture(s) in conftest.py
- adapts the tests to use pytest

schtibe commented on app/tests/conftest.py:
```diff
@@ -0,0 +1,23 @@
+    ) -> Any:
+        user = get_user_model().objects.create_user(username=username, password=password)
+        for app_label, model, codename in permissions:
+            content_type = ContentType.objects.get(app_label=app_label, model=model)
```
Maybe add a little comment here why this is necessary.

msom replied:
Done

schtibe commented on app/tests/access/test_access_api.py:
```diff
@@ -0,0 +1,609 @@
+
+
+@fixture(name='access_data')
+def fixture_access_data(db):
```
The name `access_data` confuses me a bit. I thought this data was for testing the "access", but below it seems to be about the user in the db to run the tests on

schtibe replied:
OK so it's access data because this here is the tests for the access app? Maybe add a little docstring to the fixture describing the purpose of it

schtibe commented on app/tests/conftest.py:
```diff
@@ -0,0 +1,23 @@
+
+
+@fixture
+def user_factory(db):
```
The name is a bit ambiguous: it's about the user that's used to authenticate on the API, but that isn't exactly clear. It could be understood as the users in the db, the ones that we push to cognito.

msom replied:
Renamed to django_user_factory

schtibe commented on app/tests/cognito/test_cognito_client.py:
```diff
@@ -0,0 +1,255 @@
+from django.conf import settings
+
+
+def cognito_user(username, managed, for_):
```
I think it's good practice if functions contain a verb. Unless you turn this into a fixture of course.

msom replied:
Moved to global fixture `cognito_user_response_factory`

schtibe commented on app/tests/distributions/test_attribution_model.py:
```diff
@@ -0,0 +1,130 @@
+            model = Attribution
+            fields = "__all__"
+
+    provider = Provider.objects.create()
```
I guess you could theoretically omit the `db` fixture here and create the Model via `Provider` 😅

msom replied:
moved to fixture

schtibe commented on app/tests/distributions/test_distributions_api.py:
```diff
@@ -0,0 +1,727 @@
+
+
+def test_get_attribution_returns_existing_attribution_with_default_language(
+    distribution_data, client, user_factory
```
I'm wondering if we should be more concise with the names of the fixtures. `user_factory` is called "factory" but `distribution_data` is not. To me, `distribution_data` is also sort of a factory (even if we don't call it explicitly, it's sort of being called without argument). 
Maybe that's influenced from using https://pytest-factoryboy.readthedocs.io/en/stable/ in the past, idk. 
I'm not saying that we should now call them all `distribution_data_factory` per se, I just want to trigger some re-evaluating of the names. (This probably applies to most of the tests)

schtibe replied:
OK now having seen all the tests maybe it's actually OK this way. I would consider renaming `user_factory` to make it clear that this is about the user that's connecting to the API, see my other comment.

msom replied:
changed all data fixtures to individual object fixtures

schtibe reviewed (changes requested):
Very nice, thank you!

I have made some comments along the reviewing concerning the naming of things. It wasn't always a 100% apparent what is what, maybe this gives you a hint for improvements.

msom commented:
Your right, naming was confusing! Cleanup done:
- Changed all the data fixtures to individual fixtures returning single database objects
- Renamed `user_factory` to `django_user_factory`
- Moved the functionality of `cognito_user` to a global fixture `cognito_user_response_factory`
- Documented the global (factory) fixtures

schtibe reviewed (approved):
Nice, you made it even better! Thank you!
